examples/switching_linear_dynamic_model.py

```diff
@@ -0,0 +1,231 @@
+logging.basicConfig(format='%(relativeCreated) 9d %(message)s', level=logging.INFO)
+
+@poutine.broadcast
+def model_slds(sequences, lengths, args, batch_size=None, include_prior=True):
+    num_sequences, max_length, data_dim = sequences.shape
+    assert lengths.shape == (num_sequences,)
+    assert lengths.max() <= max_length
+
+    n_seg = args.num_segment # K, number of segments
+    n_lat = args.num_latent
+    n_out = data_dim
+    normal_std = 1e-1
+    with poutine.mask(mask=torch.tensor(include_prior)):
+        # transition matrix for HMM, K*K,
+        hmm_dyn = pyro.sample("hmm_dyn",
+                              dist.Dirichlet(0.9 * torch.eye(n_seg) + 0.1)
+                                  .independent(1))
+        ## LDS
+        ssm_dyn = pyro.sample("ssm_dyn",
+                              dist.Normal(0,normal_std).expand_by([n_seg, n_lat, n_lat])
+                                  .independent(3))
+        ssm_bias = pyro.sample("ssm_bias",
+                              dist.Normal(0,1.).expand_by([n_seg, n_lat, 1])
+                                  .independent(3))
+        ssm_noise = pyro.sample("ssm_noise",
+                              dist.Gamma(1e0,1e0).expand_by([n_seg, n_lat, 1])
+                                  .independent(3))
+        ## observation
+        obs_weight = pyro.sample("obs_weight",
+                              dist.Normal(0,1).expand_by([n_seg, n_out, n_lat])
+                                  .independent(3))
+        obs_bias = pyro.sample("obs_bias",
+                              dist.Normal(0,1).expand_by([n_seg, n_out, 1])
+                                  .independent(3))
+        obs_noise = pyro.sample("obs_noise",
+                              dist.Gamma(1e0,1e0).expand_by([n_seg, n_out, 1])
+                                  .independent(3))
+
+    with pyro.iarange("sequences", len(sequences), batch_size, dim=-1) as batch:
+        ## ssm init state
+        # use a (n_lat,1) matrix (instead of vector) to ease batch operation,
+        # as for high dim tensor , matmul would do batch matrix-matrix product, instead of matrix-vector product.
+        #
+        ssm_init = pyro.sample("ssm_init", dist.Normal(0,1).expand_by([n_lat,1]).independent(2))
+
+        lengths = lengths[batch]
+        z = 0
+        x = ssm_init
+        for t in range(lengths.max()):
+            with poutine.mask(mask=(t < lengths).unsqueeze(-1)):
+                z = pyro.sample("z_{}".format(t), dist.Categorical(hmm_dyn[z]),
+                                infer={"enumerate": "parallel"})
+                x = pyro.sample("x_{}".format(t), dist.Normal(torch.matmul(ssm_dyn[z], x) + ssm_bias[z], torch.sqrt(1./ssm_noise[z])).independent(2))
+                obs_lat = torch.matmul(obs_weight[z], x) + obs_bias[z]
+                pyro.sample("y_{}".format(t), dist.Normal(obs_lat, torch.sqrt(1./obs_noise[z])).independent(2),
+                            obs=sequences[batch, t].unsqueeze(-1))
```

Comment:
I think the error may be on this line. I don't fully inderstand PyTorch indexing, but I find that to index into a tensor using tensor indices, I need to index on either a single dimensions or on all dimensions. Here you are indexing on only two of the three dimensions. I think you could instead use either
```py
pyro.sample("y_{}"..., obs=sequences[batch][:,t])
```
or
```py
data_index = torch.arange(data_dim, dtype=torch.long)
...
pyro.sample("y_{}"..., obs=sequences[batch, t, data_index])
```

Reply:
Hmm actually I think the error is elsewhere.